rust/cymbal/src/symbol_store/caching.rs

```diff
@@ -0,0 +1,121 @@
+use std::{collections::HashMap, sync::Arc, time::Instant};
+
+use axum::async_trait;
+use tokio::sync::Mutex;
+
+use crate::{
+    error::Error,
+    metric_consts::{
+        STORE_CACHED_BYTES, STORE_CACHE_EVICTIONS, STORE_CACHE_HITS, STORE_CACHE_MISSES,
+        STORE_CACHE_SIZE,
+    },
+};
+
+use super::{SymbolSetRef, SymbolStore};
+
+pub struct CachingStore {
+    inner: Box<dyn SymbolStore>,
+    cache: Mutex<CacheInner>,
+}
+
+impl CachingStore {
+    pub fn new(inner: Box<dyn SymbolStore>, max_bytes: usize) -> Self {
+        metrics::gauge!(STORE_CACHE_SIZE).set(max_bytes as f64);
+        Self {
+            inner,
+            cache: Mutex::new(CacheInner {
+                cached: Default::default(),
+                max_bytes,
+                held_bytes: 0,
+            }),
+        }
+    }
+}
+
+#[async_trait]
+impl SymbolStore for CachingStore {
+    async fn fetch(&self, team_id: i32, r: SymbolSetRef) -> Result<Arc<Vec<u8>>, Error> {
+        let mut cache = self.cache.lock().await;
```

Comment:
Does the lock get dropped when `cache` gets removed?

Reply:
The lock (a `MutexGuard<_>`) gets released when the function returns, because the variable name it's bound to goes out of scope.